ingestion_tools/scripts/data_validation/run_tests.py

```diff
@@ -10,79 +9,95 @@
+    saving history will result in longer execution time (each dataset has to be an individual `pytest` call).",
+)
 @click.option(
     "--extra-args",
-    required=False,
-    type=str,
     default=None,
+    type=str,
     help="Extra arguments to pass to pytest (in one string).",
 )
 def main(
-    local_dst: str,
-    bucket: str,
+    local_dir: str,
+    output_dir: str,
+    input_bucket: str,
     output_bucket: str,
     datasets: str | None,
-    save_history: bool,
+    multiprocessing: bool,
+    history: bool,
     extra_args: str | None,
 ):
     fs: S3Filesystem = FileSystemApi.get_fs_api(mode="s3", force_overwrite=False)
-    now = datetime.datetime.now().isoformat(sep="_", timespec="seconds").replace(":", "-")
 
     if datasets is None:
-        datasets = fs.glob(f"s3://{bucket}/*")
+        datasets = fs.glob(f"s3://{input_bucket}/*")
         datasets = [os.path.basename(dataset) for dataset in datasets if os.path.basename(dataset).isdigit()]
     else:
         datasets = datasets.split(",")
 
+    datasets = sorted(datasets, key=lambda x: int(x))
+
+    # If we are not saving history, we can run all datasets in one go (individual dataset runs don't need to be tracked)
+    if not history:
+        datasets = [",".join(datasets)]
+
     extra_args = extra_args if extra_args else ""
 
     for dataset in datasets:
+        now = datetime.datetime.now().isoformat(sep="_", timespec="seconds").replace(":", "-")
+
         # Accumulate test results here
-        localdir_raw = f"{local_dst}/{dataset}_raw/{now}"
+        localdir_raw = f"{local_dir}/{dataset}_raw/{now}"
         # Generate report here
-        localdir_rep = f"{local_dst}/{dataset}/{now}"
-        # Archive location
-        local_tar = f"{local_dst}/{dataset}/{dataset}_{now}.tar.gz"
-
-        os.makedirs(localdir_raw, exist_ok=True)
-        os.makedirs(localdir_rep, exist_ok=True)
+        localdir_rep = f"{local_dir}/{dataset}/{now}"
 
         # Run tests and generate results
-        os.system(f"pytest -n auto --dist worksteal --dataset {dataset} --alluredir {localdir_raw} {extra_args}")
+        os.system(
+            f"pytest {'--dist worksteal -n auto' if multiprocessing else '--dist no'} --datasets {dataset} --alluredir {localdir_raw} {extra_args}",
+        )
 
         # Get the history from S3 (Must do this before generating the report)
-        remote_dataset_dir = f"{output_bucket}/data_validation/{dataset}"
-        remote_dataset_tar = f"{remote_dataset_dir}/{dataset}.tar.gz"
-        fs.makedirs(remote_dataset_dir)
-
-        if fs.exists(remote_dataset_tar):
-            get_history(remote_dataset_tar, localdir_raw, fs)
+        remote_dataset_dir = f"{output_bucket}/{output_dir}/{dataset}"
+        if fs.exists(remote_dataset_dir):
+            get_history(remote_dataset_dir, localdir_raw, fs)
+            fs.s3fs.rm(remote_dataset_dir, recursive=True)
```

Comment:
As there is a timestamp associated to it, what would be a usecase for this to already exist?

Reply:
To keep track of history, we currently just upload the latest report run (with the history/ folder in it) for a given dataset. We don't track every single report, but instead just the latest report (which holds the history of the last 20 reports before it). Do we want to keep every report that is run?